File: src/push.py
# push.py
#
# Copyright 2026 Unknown
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import os

from gi.repository import GLib

from .webdav import WebDAVClient
from .db import get_db


class PushEngine:
    """Upload/mirror engine that walks a local directory and pushes
    new or changed files to a remote WebDAV target."""

    def __init__(self, client, target, on_progress=None, on_complete=None):
        """
        client: WebDAVClient instance
        target: dict from db.get_target() with keys: id, local_path, remote_path, mode
        on_progress: callback(current_file, files_done, files_total) - called after each file
        on_complete: callback(success, summary_dict) - called when push finishes
            summary_dict: {'uploaded': int, 'skipped': int, 'deleted': int, 'errors': list[str]}
        """
        self._client = client
        self._target = target
        self._on_progress = on_progress
        self._on_complete = on_complete

        self._cancelled = False
        self._upload_queue = []
        self._local_rel_paths = set()
        self._files_total = 0
        self._files_done = 0
        self._summary = {
            'uploaded': 0,
            'skipped': 0,
            'deleted': 0,
            'errors': [],
        }

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def start(self, force=False):
        """Begin the push process.

        If force is True, clear all file_state records for this target
        and re-upload everything regardless of mtime/size.
        """
        self._cancelled = False
        self._summary = {
            'uploaded': 0,
            'skipped': 0,
            'deleted': 0,
            'errors': [],
        }

        db = get_db()
        target_id = self._target['id']
        local_path = self._target['local_path']
        remote_base = self._target['remote_path'].rstrip('/')

        if force:
            db.clear_file_states(target_id)

        # Synchronously walk the local directory
        all_files = []
        for dirpath, _dirnames, filenames in os.walk(local_path):
            for fname in filenames:
                abs_path = os.path.join(dirpath, fname)
                rel_path = os.path.relpath(abs_path, local_path)
                # Normalise to forward slashes for remote paths
                rel_path = rel_path.replace(os.sep, '/')
                mtime = os.path.getmtime(abs_path)
                size = os.path.getsize(abs_path)
                all_files.append({
                    'rel_path': rel_path,
                    'local_path': abs_path,
                    'remote_path': remote_base + '/' + rel_path,
                    'mtime': mtime,
                    'size': size,
                })

        # Keep a set of all local relative paths for the mirror pass
        self._local_rel_paths = {f['rel_path'] for f in all_files}

        # Filter to files that actually need uploading
        upload_list = []
        for entry in all_files:
            if force:
                upload_list.append(entry)
                continue
            state = db.get_file_state(target_id, entry['rel_path'])
            if state and state['mtime'] == entry['mtime'] and state['size'] == entry['size']:
                self._summary['skipped'] += 1
            else:
                upload_list.append(entry)

        self._upload_queue = upload_list
        self._files_total = len(upload_list)
        self._files_done = 0

        # Collect unique parent directories, sorted shallowest first
        dirs_set = set()
        for entry in upload_list:
            parent = os.path.dirname(entry['rel_path'])
            while parent:
                dirs_set.add(parent)
                parent = os.path.dirname(parent)

        dirs_sorted = sorted(dirs_set, key=lambda d: d.count('/'))
        remote_dirs = [remote_base + '/' + d for d in dirs_sorted]

        # Start the async chain: create dirs → upload files → mirror → complete
        self._create_directories(remote_dirs, self._upload_next)

    def cancel(self):
        """Cancel the running push. The upload loop will stop between files."""
        self._cancelled = True

    # ------------------------------------------------------------------
    # Internal async chain
    # ------------------------------------------------------------------

    def _create_directories(self, dirs_list, callback):
        """Create remote directories one by one via callback chaining."""
        if not dirs_list:
            callback()
            return

        remaining = list(dirs_list)

        def _create_next():
            if not remaining:
                callback()
                return
            current_dir = remaining.pop(0)

            def _on_done(success, error):
                if not success:
                    self._summary['errors'].append(
                        f"mkdir {current_dir}: {error}"
                    )
                _create_next()

            self._client.create_directory(current_dir, _on_done)

        _create_next()

    def _upload_next(self):
        """Upload the next file in the queue, chaining via WebDAV callback."""
        if self._cancelled or not self._upload_queue:
            self._on_uploads_finished()
            return

        entry = self._upload_queue.pop(0)

        def _on_upload_done(success, error):
            if success:
                db = get_db()
                db.upsert_file_state(
                    self._target['id'],
                    entry['rel_path'],
                    entry['mtime'],
                    entry['size'],
                )
                self._summary['uploaded'] += 1
                self._files_done += 1
                if self._on_progress:
                    GLib.idle_add(
                        self._on_progress,
                        entry['rel_path'],
                        self._files_done,
                        self._files_total,
                    )
            else:
                self._summary['errors'].append(
                    f"upload {entry['rel_path']}: {error}"
                )
                self._files_done += 1

            self._upload_next()

        self._client.upload_file(
            entry['local_path'], entry['remote_path'], _on_upload_done
        )

    def _on_uploads_finished(self):
        """Called after all uploads complete (or cancel). Run mirror if needed."""
        if self._target['mode'] == 'mirror' and not self._cancelled:
            self._mirror_pass(self._finish)
        else:
            self._finish()

    def _mirror_pass(self, callback):
        """Delete remote files that no longer exist locally (mirror mode).

        Lists the remote directory recursively by collecting subdirectory
        listings, then deletes any remote file whose relative path is not
        in the local file set.
        """
        remote_base = self._target['remote_path'].rstrip('/')
        to_delete = []

        # We collect all remote relative paths via recursive listing,
        # then delete those not present locally.
        def _list_recursive(remote_dir, prefix, done_cb):
            """List *remote_dir*, recurse into subdirectories, call done_cb
            when the full subtree has been enumerated."""
            self._client.list_directory(remote_dir, lambda ok, res: _on_list(
                ok, res, remote_dir, prefix, done_cb,
            ))

        def _on_list(success, result, remote_dir, prefix, done_cb):
            if not success:
                self._summary['errors'].append(
                    f"mirror list {remote_dir}: {result}"
                )
                done_cb()
                return

            # Separate files and subdirectories
            subdirs = []
            for item in result:
                rel = prefix + '/' + item['name'] if prefix else item['name']
                if item['is_dir']:
                    subdirs.append((remote_dir.rstrip('/') + '/' + item['name'], rel))
                else:
                    if rel not in self._local_rel_paths:
                        to_delete.append(remote_base + '/' + rel)

            # Recurse into subdirectories sequentially
            def _recurse_subdirs():
                if not subdirs:
                    done_cb()
                    return
                rd, rp = subdirs.pop(0)
                _list_recursive(rd, rp, _recurse_subdirs)

            _recurse_subdirs()

        def _on_listing_complete():
            # Delete collected remote paths one by one
            self._delete_remote_files(to_delete, callback)

        _list_recursive(remote_base, '', _on_listing_complete)

    def _delete_remote_files(self, paths, callback):
        """Delete a list of remote paths sequentially via callback chaining."""
        if not paths:
            callback()
            return

        remaining = list(paths)

        def _delete_next():
            if not remaining:
                callback()
                return
            current = remaining.pop(0)

            def _on_deleted(success, error):
                if success:
                    self._summary['deleted'] += 1
                else:
                    self._summary['errors'].append(
                        f"delete {current}: {error}"
                    )
                _delete_next()

            self._client.delete(current, _on_deleted)

        _delete_next()

    def _finish(self):
        """Final step: update DB timestamp and fire the on_complete callback."""
        db = get_db()
        db.update_last_push(self._target['id'])

        success = not self._summary['errors']
        if self._on_complete:
            GLib.idle_add(self._on_complete, success, self._summary)

File: src/push.rs
use crate::db::Target;
use crate::webdav::WebDAVClient;
use std::collections::HashSet;
use std::path::Path;
use std::sync::mpsc;

#[derive(Debug, Clone, Default)]
pub struct PushSummary {
    pub uploaded: u32,
    pub skipped: u32,
    pub deleted: u32,
    pub errors: Vec<String>,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub enum PushProgress {
    File {
        current_file: String,
        files_done: u32,
        files_total: u32,
    },
    Complete {
        success: bool,
        summary: PushSummary,
    },
}

/// Walk local directory and collect files with metadata
fn collect_local_files(local_path: &str) -> Result<Vec<(String, f64, i64)>, String> {
    let base = Path::new(local_path);
    if !base.is_dir() {
        return Err(format!("Local path does not exist: {}", local_path));
    }

    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(base)
        .follow_links(false)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        if entry.file_type().is_file() {
            let abs_path = entry.path();
            let rel_path = abs_path.strip_prefix(base)
                .map_err(|e| e.to_string())?
                .to_string_lossy()
                .to_string();

            let metadata = abs_path.metadata().map_err(|e| e.to_string())?;
            let mtime = metadata.modified()
                .map_err(|e| e.to_string())?
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_secs_f64())
                .unwrap_or(0.0);
            let size = metadata.len() as i64;

            files.push((rel_path, mtime, size));
        }
    }

    Ok(files)
}

/// Run a push operation for a single target on a background thread.
/// Sends progress updates via the provided sender.
pub fn run_push(
    client: WebDAVClient,
    target: Target,
    db_path: std::path::PathBuf,
    force: bool,
    sender: mpsc::Sender<PushProgress>,
) {
    std::thread::spawn(move || {
        let result = do_push(&client, &target, &db_path, force, &sender);
        match result {
            Ok(summary) => {
                let _ = sender.send(PushProgress::Complete {
                    success: summary.errors.is_empty(),
                    summary,
                });
            }
            Err(e) => {
                let _ = sender.send(PushProgress::Complete {
                    success: false,
                    summary: PushSummary {
                        errors: vec![e],
                        ..Default::default()
                    },
                });
            }
        }
    });
}

fn do_push(
    client: &WebDAVClient,
    target: &Target,
    db_path: &Path,
    force: bool,
    sender: &mpsc::Sender<PushProgress>,
) -> Result<PushSummary, String> {
    // Open a dedicated DB connection for this thread
    let conn = rusqlite::Connection::open(db_path)
        .map_err(|e| format!("DB error: {}", e))?;
    conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA foreign_keys=ON;").ok();

    let mut summary = PushSummary::default();

    // Step 1: Collect local files
    let local_files = collect_local_files(&target.local_path)?;
    let _files_total = local_files.len() as u32;

    if force {
        conn.execute("DELETE FROM file_state WHERE target_id = ?1",
            rusqlite::params![target.id]).ok();
    }

    // Step 2: Filter files that need uploading (changed since last push)
    let mut to_upload: Vec<(String, f64, i64)> = Vec::new();
    for (rel_path, mtime, size) in &local_files {
        let existing: Option<(f64, i64)> = conn.query_row(
            "SELECT mtime, size FROM file_state WHERE target_id = ?1 AND rel_path = ?2",
            rusqlite::params![target.id, rel_path],
            |row| Ok((row.get(0)?, row.get(1)?)),
        ).ok();

        if let Some((old_mtime, old_size)) = existing {
            if (old_mtime - mtime).abs() < 0.001 && old_size == *size {
                summary.skipped += 1;
                continue;
            }
        }
        to_upload.push((rel_path.clone(), *mtime, *size));
    }

    // Step 3: Create remote directories
    let mut dirs: Vec<String> = to_upload.iter()
        .filter_map(|(rel, _, _)| {
            Path::new(rel).parent().map(|p| p.to_string_lossy().to_string())
        })
        .filter(|d| !d.is_empty())
        .collect::<HashSet<_>>()
        .into_iter()
        .collect();
    dirs.sort();

    for dir in &dirs {
        let remote_dir = format!("{}/{}", target.remote_path.trim_end_matches('/'), dir);
        // Create parent directories from top to bottom
        let parts: Vec<&str> = remote_dir.trim_start_matches('/').split('/').collect();
        let mut path_so_far = String::new();
        for part in parts {
            path_so_far = if path_so_far.is_empty() {
                part.to_string()
            } else {
                format!("{}/{}", path_so_far, part)
            };
            if let Err(e) = client.create_directory(&path_so_far) {
                summary.errors.push(format!("mkdir {}: {}", path_so_far, e));
            }
        }
    }

    // Step 4: Upload files
    let upload_total = to_upload.len() as u32;
    for (i, (rel_path, mtime, size)) in to_upload.iter().enumerate() {
        let local_file = format!("{}/{}", target.local_path, rel_path);
        let remote_file = format!("{}/{}", target.remote_path.trim_end_matches('/'), rel_path);

        let _ = sender.send(PushProgress::File {
            current_file: rel_path.clone(),
            files_done: i as u32,
            files_total: upload_total,
        });

        match client.upload_file(&local_file, &remote_file) {
            Ok(()) => {
                summary.uploaded += 1;
                // Record file state
                let now = chrono::Local::now().format("%Y-%m-%dT%H:%M:%S").to_string();
                conn.execute(
                    "INSERT INTO file_state (target_id, rel_path, mtime, size, uploaded_at)
                     VALUES (?1, ?2, ?3, ?4, ?5)
                     ON CONFLICT(target_id, rel_path) DO UPDATE SET mtime = ?3, size = ?4, uploaded_at = ?5",
                    rusqlite::params![target.id, rel_path, mtime, size, now],
                ).ok();
            }
            Err(e) => {
                summary.errors.push(format!("{}: {}", rel_path, e));
            }
        }
    }

    // Step 5: Mirror mode - delete remote files not in local
    if target.mode == "mirror" {
        match client.list_directory_recursive(&target.remote_path) {
            Ok(remote_files) => {
                let local_set: HashSet<String> = local_files.iter()
                    .map(|(rel, _, _)| {
                        format!("{}/{}", target.remote_path.trim_end_matches('/'), rel)
                    })
                    .collect();

                for remote_file in remote_files {
                    if !local_set.contains(&remote_file) {
                        match client.delete(&remote_file) {
                            Ok(()) => summary.deleted += 1,
                            Err(e) => summary.errors.push(format!("delete {}: {}", remote_file, e)),
                        }
                    }
                }
            }
            Err(e) => {
                summary.errors.push(format!("Mirror listing failed: {}", e));
            }
        }
    }

    // Step 6: Update last_push
    let now = chrono::Local::now().format("%Y-%m-%dT%H:%M:%S").to_string();
    conn.execute(
        "UPDATE targets SET last_push = ?1 WHERE id = ?2",
        rusqlite::params![now, target.id],
    ).ok();

    Ok(summary)
}
